ui/org.eclipse.pde.genericeditor.extension/src/org/eclipse/pde/internal/genericeditor/target/extension/model/RepositoryCache.java

```diff
@@ -31,41 +42,73 @@
  */
 public class RepositoryCache {
 
-	private static RepositoryCache instance;
-
-	Map<String, List<UnitNode>> cache = new HashMap<>();
-
 	private RepositoryCache() {
 		//avoid instantiation
 	}
 
-	/**
-	 * @return default instance of this cache.
-	 */
-
-	public static RepositoryCache getDefault() {
-		if (instance == null) {
-			instance = new RepositoryCache();
-		}
-		return instance;
-	}
+	private static final Map<String, CompletableFuture<Map<String, List<IVersionedId>>>> CACHE = new ConcurrentHashMap<>();
 
 	/**
 	 * Fetches information and caches it.
-	 *
-	 * @param repo
-	 *            repository URL
-	 * @param flush
-	 *            whether a flush is needed
-	 * @return list of IUs available in the 'repo' repository. Never
-	 *         <code>null</code>.
+	 * <p>
+	 * All available IUs are returned as a map mapping the IDs of all IUs
+	 * available in the {@code repositories} to the list of all available
+	 * {@link IVersionedId versioned IDs} for that ID. All keys are sorted in
+	 * alphabetical order and all versions are sorted in descending order.
+	 * </p>
+	 * 
+	 * @return all available units in the specified {@code repository} in a map
+	 *         mapping all IDs to all available versions.
 	 */
-	public List<UnitNode> fetchP2UnitsFromRepo(String repo, boolean flush) {
-		if ((flush) || (cache.get(repo) == null)) {
-			List<UnitNode> units = P2Fetcher.fetchAvailableUnits(repo);
-			cache.put(repo, units);
+	public static Map<String, List<IVersionedId>> fetchP2UnitsFromRepos(List<String> repositories) {
+		if (repositories.size() == 1) {
+			try {
+				return fetchP2DataOfRepo(repositories.get(0)).get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
 		}
-		return cache.get(repo);
+		List<Future<Map<String, List<IVersionedId>>>> repos = repositories.stream()
+				.map(RepositoryCache::fetchP2DataOfRepo).toList();
+		// Fetch all repos at once to await pending metadata in parallel
+		return toSortedMap(repos.stream().<Map<String, List<IVersionedId>>>map(f -> {
+			try {
+				return f.get();
+			} catch (InterruptedException | ExecutionException e) {
+				return Map.of();
+			}
+		}).map(Map::values).flatMap(Collection::stream).flatMap(List::stream));
+	}
+
+	public static void prefetchP2MetadataOfRepository(String repository) {
+		fetchP2DataOfRepo(repository);
+	}
+
+	private static Future<Map<String, List<IVersionedId>>> fetchP2DataOfRepo(String repository) {
+		return CACHE.compute(repository, (r, f) -> {
+			if (f != null && f.isDone() && !f.isCompletedExceptionally() && !f.isCancelled()) {
+				return f;
+			}
+			CompletableFuture<Map<String, List<IVersionedId>>> future = new CompletableFuture<>();
+			// Fetching P2 repository information is a costly operation
+			// time-wise. Thus it is done in a job.
+			Job job = Job.create(NLS.bind(Messages.UpdateJob_P2DataFetch, r), m -> {
+				Map<String, List<IVersionedId>> map = toSortedMap(P2Fetcher.fetchAvailableUnits(r));
```

Comment:
Maybe pass the monitor to P2Fetcher so it could give some progress there...

Reply:
> Maybe pass the monitor to P2Fetcher so it could give some progress there...

Yes. I have noticed that as well and started to add it locally.

> 
> Beside that one should catch any runtime exceptions here and complete the future exceptionally, also P2 fetcher seem to catch and ignore some exceptions already so a future will never be finished with an error.

Yes the exceptional completion of the Future was indeed missing. Added it. But in this case it does not really make a difference if the job was canceled or another exception occurred. Therefore I just complete exceptionally in case of any exception, regardless of if it's a CancellationException.

`P2Fechter` catches and rethrows some exception, only the exception when creating the agent was swallowed, but lead to a subsequent NPE.
But I have now cleaned all that up to not catch any `CoreException` and just propagate all of them.

> 
> Also if error or cancelation occurs, it might be good to remove the future from the map to not retain it forever in the cache until it is queried next.

Don't think a failed future consumes much memory? And removing it could maybe lead to subtitle concurrency issues.
So I would just keep it simple and clean it up on the next content assist interaction.